### PR TITLE
feat(pt): Add DPA4/SeZM descriptor & model 🎉🎉🎉

### PR DESCRIPTION
**Pull request by OutisLi**
## Summary

This PR adds PyTorch support for DPA4, the DeePMD-kit implementation of SeZM (Smooth Equivariant Zone-bridging Model). It introduces the DPA4/SeZM model, descriptor, fitting network, training integration, export path, documentation, examples, and tests.

## Main Changes

- Add the DPA4/SeZM PyTorch model stack:
  - `model.type: "dpa4"` / `"sezm"`
  - `descriptor.type: "dpa4"` / `"sezm"`
  - `fitting_net.type: "dpa4_ener"` / `"sezm_ener"`
- Implement the SO(3)-equivariant descriptor with edge-local SO(2) convolutions, angular schedules, smooth radial envelopes, attention/focus streams, and environment-seeded initial features.
- Add zone-bridging support for short-range analytical repulsion, including ZBL coupling and descriptor-side short-range clamping.
- Add DPA4 training support for:
  - conservative energy/force training through `loss.type: "ener"`
  - experimental direct-force denoising through `loss.type: "dens"`
  - spin models in the PyTorch backend
  - shared-fitting multitask case FiLM conditioning
  - LoRA fine-tuning and merged checkpoint export
- Add the DPA4 `.pt2` freeze/export path using AOTInductor for checkpoints that cannot be represented by the regular TorchScript freeze path.
- Add CLI, argcheck, validation, data-system, and inference integration needed to route DPA4 configs and exported models correctly.
- Add water examples for standard DPA4, ZBL bridging, spin, DeNS, multitask/shared-fitting, LoRA fine-tuning, and LAMMPS inference.
- Add official model documentation at `doc/model/dpa4.md`.

## Tests

This PR adds coverage for:

- DPA4/SeZM model and descriptor construction
- DPA4 aliases in model, descriptor, and fitting configuration
- SO(3)/SO(2) equivariance behavior
- conservative energy/force paths
- `torch.compile` eager/compiled consistency
- DPA4 `.pt2` export and DeepPot inference
- spin model behavior
- ZBL zone bridging
- DeNS loss and direct-force mode
- LoRA adapter injection, freezing, merging, and compile compatibility
- optional Triton kernel dispatch and numerical consistency
- supporting utility changes in neighbor-list, LMDB data, and distributed checks

Relevant test files include:

- `source/tests/pt/model/test_descriptor_sezm.py`
- `source/tests/pt/model/test_descriptor_sezm_s2_equivariance.py`
- `source/tests/pt/model/test_descriptor_sezm_triton.py`
- `source/tests/pt/model/test_sezm_model.py`
- `source/tests/pt/model/test_sezm_spin_model.py`
- `source/tests/pt/model/test_sezm_export.py`
- `source/tests/pt/test_training.py`
- `source/tests/pt/test_train_utils.py`
- `source/tests/common/dpmodel/test_dist_check.py`
- `source/tests/common/dpmodel/test_lmdb_data.py`

## Notes

DPA4 is currently implemented for the PyTorch backend. Model compression is not supported, and DPA4 checkpoints use the `.pt2` export path instead of the regular TorchScript freeze path.

## Summary by CodeRabbit

* **New Features**
  * Added SeZM model family and DeNS denoising loss for training; new optimized ".pt2" export path with embedded metadata.

* **Improvements**
  * LoRA fine-tuning workflow (apply/merge/strip) for lightweight adapters.
  * On-demand minimum pairwise-distance computation during data reads.
  * Better JAX neighbor-list handling and optional GPU/Triton-accelerated descriptor kernels for faster inference/training.

[Review Change Stack](https://app.coderabbit.ai/change-stack/deepmodeling/deepmd-kit/pull/5448?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: deepmd/pt/model/descriptor/sezm_nn/so2.py [hunk 1]
```diff
@@ -0,0 +1,1624 @@
...
+                        scale = self.adam_so2_layer_scales[layer_idx].reshape(
+                            1, self.n_focus, 1, self.so2_focus_dim
+                        )
+                        x_local = residual + scale * x_local
+                    else:
+                        x_local = residual + x_local
+
+        # === Step 6. Cross-focus softmax competition ===
+        if self.focus_compete and self.n_focus > 1:
+            focus_gate_src = focus_gate_src.to(dtype=self.compute_dtype)
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt/model/descriptor/sezm_nn/so2.py:1425]: if self.focus_compete and self.n_focus > 1:
                focus_gate_src = x_local[:, :, 0, :]

File: deepmd/pt/model/descriptor/sezm_nn/triton/constants.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+_HAS_TORCH_TRITON_OP = hasattr(torch.library, "triton_op") and hasattr(
+    torch.library, "wrap_triton"
+)
+
+if _HAS_TORCH_TRITON_OP:
+    try:
+        import triton  # noqa: F401
+    except ImportError:
+        SEZM_TRITON_AVAILABLE = False
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt/model/descriptor/sezm_nn/triton/constants.py:22]: Used

File: examples/water/dpa4/input_multitask_sharefit.json
Comment (njzjz) [inline examples/water/dpa4/input_multitask_sharefit.json:1]: Please add these files to `test_examples.py`, which is used to ensure the example files are correct and up-to-date.

File: deepmd/pt/model/descriptor/sezm.py [hunk 3]
```diff
@@ -0,0 +1,1951 @@
...
+            type_ebed = self.type_embedding(atype_loc).reshape(
+                -1, self.channels
+            )  # (N, C)
+            if self.charge_spin_embedding is not None:
+                type_ebed = self._apply_charge_spin_embedding(
+                    type_ebed,
+                    charge_spin,
+                    nf=nf,
+                    nloc=nloc,
+                )
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt/model/descriptor/sezm.py:1179]: done in sezm_model.py through convert_charge_spin()

File: examples/water/dpa4/lmp/README.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+```bash
+dp --pt train input.json --skip-neighbor-stat
+```
+
+Freeze the checkpoint (the pt backend detects DPA4 / SeZM and writes a
+`.pt2` archive automatically):
+
+```bash
+dp --pt freeze -c model.ckpt.pt -o frozen_model
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply examples/water/dpa4/lmp/README.md:32]: I want to keep this

File: examples/water/dpa4/lmp/.gitignore
Comment (iProzd) [inline examples/water/dpa4/lmp/.gitignore:1]: this file should be removed.

File: examples/water/dpa4/.gitignore
Comment (iProzd) [inline examples/water/dpa4/.gitignore:1]: this file should be removed.

**Review by iProzd — CHANGES_REQUESTED** [review]

**Comment by iProzd** [comment]
Others LGTM.

File: deepmd/pt/utils/multi_task.py [hunk 5]
```diff
@@ -14,6 +14,26 @@
 )
 
 
+def _cascade_top_level_defaults(model_config: dict[str, Any]) -> None:
```
Comment (wanghan-iapcm) [inline deepmd/pt/utils/multi_task.py:17]: recommend adding a test for the function covering:1 
  1. Cascade-on-missing — top-level key lands in a branch that omitted it.                                                                                                                     
  2. Explicit-wins precedence — top-level value does not clobber a branch that set the key.
  3. Pop from top level — after preprocessing, the top-level key is gone (so argcheck doesn't reject it).                                                                                      

```python
  def test_cascade_top_level_defaults():                                                                                                                                                       
      # 1. cascade-on-missing                                              
      cfg = {"foo": 1, "model_dict": {"a": {}, "b": {"foo": 2}}}                                                                                                                               
      _cascade_top_level_defaults(cfg)                                                                                                                                                         
      assert cfg["model_dict"]["a"]["foo"] == 1                                                                                                                                                
      # 2. explicit-wins                                                                                                                                                                       
      assert cfg["model_dict"]["b"]["foo"] == 2                                                                                                                                              
      # 3. popped from top level                                                                                                                                                               
      assert "foo" not in cfg                                              
      # 4. shared_dict / model_dict are not cascaded                                                                                                                                           
      cfg2 = {"shared_dict": {"x": 1}, "model_dict": {"a": {}}}                                        
      _cascade_top_level_defaults(cfg2)                                                                                                                                                        
      assert "shared_dict" in cfg2 and "shared_dict" not in cfg2["model_dict"]["a"]                    
      # 5. deepcopy independence (mutating one branch's value doesn't bleed)                                                                                                                   
      cfg3 = {"foo": [1, 2], "model_dict": {"a": {}, "b": {}}}                                                                                                                                 
      _cascade_top_level_defaults(cfg3)                                                                                                                                                        
      cfg3["model_dict"]["a"]["foo"].append(99)                                                                                                                                                
      assert cfg3["model_dict"]["b"]["foo"] == [1, 2]                                                                                                                                        
```

File: deepmd/pt/model/atomic_model/sezm_atomic_model.py [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+            Serialized atomic-model data.
+
+        Returns
+        -------
+        SeZMAtomicModel
+            Deserialized SeZM atomic model.
+        """
+        payload = data.copy()
+        version = int(payload.pop("@version", 2))
+        if version not in (2, 3):
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/atomic_model/sezm_atomic_model.py]: perhaps using `check_version_compatibility` is better for maintenance.

File: deepmd/dpmodel/utils/dist_check.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    ----------
+    coord : np.ndarray
+        Atomic coordinates, flattened with shape (natoms * 3,)
+        or reshaped as (natoms, 3).
+    box : np.ndarray or None
+        Box vectors with shape (9,) for PBC, or None for non-PBC.
+    atype : np.ndarray
+        Atom types with shape (natoms,). Virtual atoms (type < 0)
+        are excluded from the distance check.
+    stop_below : float or None
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/utils/dist_check.py:31]: this argument is not well tested by UT

```python
  def test_stop_below_triggers_early_exit(self):
      # Coordinates where the first block already contains a pair < stop_below
      coord = np.array([0., 0., 0., 0.05, 0., 0., 10., 0., 0.])                                                                                                                                
      atype = np.array([0, 0, 0])                                                                                                                                                              
      # With stop_below=0.1, the 0.05 pair should trigger early exit; result still correct                                                                                                     
      d = compute_min_pair_dist_single(coord, box=None, atype=atype, stop_below=0.1)                                                                                                           
      np.testing.assert_almost_equal(d, 0.05)                                                                                                                                                  
                                                                                                                                                                                               
  def test_stop_below_not_triggered(self):                                                                                                                                                     
      # All pairs >= stop_below; result is the true minimum                                                                                                                                  
      coord = np.array([0., 0., 0., 1.0, 0., 0., 2.0, 0., 0.])                                                                                                                                 
      atype = np.array([0, 0, 0])                                                                                                                                                              
      d = compute_min_pair_dist_single(coord, box=None, atype=atype, stop_below=0.5)                                                                                                           
      np.testing.assert_almost_equal(d, 1.0)                                                                                                                                                   
                                                                                                                                                                                             
  def test_multi_block_iteration(self):                                                                                                                                                        
      # > 512 atoms forces multi-block path                                
      rng = np.random.default_rng(42)                                                                                                                                                          
      n = 600
      coord = rng.uniform(0, 100, (n, 3))                                                                                                                                                      
      atype = np.zeros(n, dtype=np.int64)                                  
      # Reference: brute-force pairwise distance min                                                                                                                                           
      diff = coord[:, None] - coord[None, :]
      ref = np.sqrt((diff * diff).sum(-1) + np.eye(n) * np.inf).min()                                                                                                                          
      d = compute_min_pair_dist_single(coord, box=None, atype=atype)                                                                                                                           
      np.testing.assert_almost_equal(d, ref, decimal=10)                                                                                                                                       
                                                                                                                                                                                               
  def test_coincident_atoms_zero(self):                                                                                                                                                      
      coord = np.array([0., 0., 0., 0., 0., 0., 1., 0., 0.])                                                                                                                                 
      atype = np.array([0, 0, 0])
      d = compute_min_pair_dist_single(coord, box=None, atype=atype)
      self.assertEqual(d, 0.0)                                                                                                                                                                 
```

File: deepmd/dpmodel/utils/nlist.py [hunk 8]
```diff
@@ -355,9 +355,17 @@ def extend_coord_with_ghosts(
         shift_idx = xp.take(xyz, xp.argsort(xp.linalg.vector_norm(xyz, axis=1)), axis=0)
         ns, _ = shift_idx.shape
         nall = ns * nloc
-        # shift_vec = xp.einsum("sd,fdk->fsk", shift_idx, cell)
-        shift_vec = xp.tensordot(shift_idx, cell, axes=([1], [1]))
-        shift_vec = xp.permute_dims(shift_vec, (1, 0, 2))
+        xp_name = getattr(xp, "__name__", "")
+        if "jax" in xp_name:
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/utils/nlist.py]: use `is_jax_namespace(xp) `

File: pyproject.toml [hunk 9]
```diff
@@ -51,12 +51,14 @@ dependencies = [
...
     "h5py>=3.6.0,!=3.11.0; platform_system=='Linux' and platform_machine=='aarch64'",
     'wcmatch',
+    "einops",
     'packaging',
     'ml_dtypes',
     'mendeleev',
     'array-api-compat',
     'lmdb',
     'msgpack',
+    'e3nn>=0.5.9',
```
Comment (njzjz) [inline pyproject.toml]: This dependency should be added to `torch` extra instead of here. It seems to be a torch-only dependency.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/dpmodel/utils/nlist.py [hunk 10]
```diff
@@ -355,9 +355,17 @@ def extend_coord_with_ghosts(
...
+            # Avoid JAX internal errors in tensordot.
+            shift_vec = xp.sum(
+                shift_idx[xp.newaxis, :, :, xp.newaxis] * cell[:, xp.newaxis, :, :],
+                axis=2,
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/utils/nlist.py:362]: the jax path should be tested by UT.

File: deepmd/pt/train/utils.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+import torch
+
+if TYPE_CHECKING:
+    from collections.abc import (
+        Callable,
+        Iterable,
+    )
+
+
+def clip_grad_norm_with_stable_fallback(
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/utils.py:22]: additional tests improve the test coverage: 
```python
  def test_fsdp_path_finite_grads(self):                                                                                                                                                       
      # use_stable_fallback=False, finite grads → returns native norm      
      p = torch.nn.Parameter(torch.zeros(1)); p.grad = torch.tensor([2.0])                                                                                                                     
      norm = clip_grad_norm_with_stable_fallback(
          [p], max_norm=1.0, use_stable_fallback=False,                                                                                                                                        
          named_parameters=lambda: [("p", p)],                             
      )                                                                                                                                                                                        
      self.assertTrue(torch.isfinite(norm))                                
      self.assertAlmostEqual(p.grad.item(), 1.0, places=5)                                                                                                                                     
                                                                                                                                                                                               
  def test_fsdp_path_nonfinite_raises(self):                                                                                                                                                 
      # use_stable_fallback=False, NaN grad → raise with named param                                                                                                                           
      p = torch.nn.Parameter(torch.zeros(1)); p.grad = torch.tensor([float("nan")])                                                                                                            
      with self.assertRaisesRegex(RuntimeError, "p:"):                                                                                                                                       
          clip_grad_norm_with_stable_fallback(                                                                                                                                                 
              [p], max_norm=1.0, use_stable_fallback=False,                
              named_parameters=lambda: [("p", p)],                                                                                                                                             
          )                                                                
                                                                                                                                                                                               
  def test_stable_fallback_nan_individual_grad_raises(self):                                                                                                                                   
      # use_stable_fallback=True, one grad is NaN → stable_clip_grad_norm raises with name                                                                                                   
      p0 = torch.nn.Parameter(torch.zeros(1)); p0.grad = torch.tensor([float("nan")])                                                                                                          
      p1 = torch.nn.Parameter(torch.zeros(1)); p1.grad = torch.tensor([1.0])                                                                                                                   
      with self.assertRaisesRegex(RuntimeError, "p0:"):                                                                                                                                        
          clip_grad_norm_with_stable_fallback(                                                                                                                                                 
              [p0, p1], max_norm=1.0,                                                                                                                                                          
              named_parameters=lambda: [("p0", p0), ("p1", p1)],                                                                                                                             
          )                                                                                                                                                                                    
                                                                           
  def test_healthy_path_no_overflow(self):                                                                                                                                                     
      # finite grads, no overflow → native path, no fallback
      p = torch.nn.Parameter(torch.zeros(1)); p.grad = torch.tensor([0.5])                                                                                                                     
      norm = clip_grad_norm_with_stable_fallback(                                                                                                                                              
          [p], max_norm=1.0,                                                                                                                                                                 
          named_parameters=lambda: [("p", p)],                                                                                                                                                 
      )                                                                    
      self.assertTrue(torch.isfinite(norm))                                                                                                                                                    
      self.assertAlmostEqual(p.grad.item(), 0.5, places=5)  # below max_norm, not clipped
                                                                                                                                                                                               
  def test_empty_parameters(self):
      norm = clip_grad_norm_with_stable_fallback([], max_norm=1.0)                                                                                                                             
      self.assertEqual(norm.item(), 0.0)                                                                                                                                                       

```

File: source/api_cc/src/DeepPotPTExpt.cc [hunk 12]
```diff
@@ -126,6 +131,23 @@ void DeepPotPTExpt::init(const std::string& model,
...
+    for (const auto& v : metadata["default_chg_spin"].as_array()) {
+      default_chg_spin_.push_back(v.as_double());
+    }
+    if (static_cast<int>(default_chg_spin_.size()) != dim_chg_spin) {
+      throw deepmd::deepmd_exception("default_chg_spin length (" +
+                                     std::to_string(default_chg_spin_.size()) +
+                                     ") does not match dim_chg_spin (" +
+                                     std::to_string(dim_chg_spin) + ").");
+    }
+  }
```
Comment (wanghan-iapcm) [inline source/api_cc/src/DeepPotPTExpt.cc]: this path is not covered by UT.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

**Review by njzjz — APPROVED** [review]
